p4_symbolic/ir/ir.proto

```diff
@@ -0,0 +1,230 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This file defines the overall structure of a P4 program in our IR, as well
+// as the supported statements and expressions.
+// The structure is tailored to be easy to consume rather than produce.
+// It deviates from the underlying bmv2 JSON format, which is mimiced closely
+// by p4_symbolic/bmv2/bmv2.proto.
+// The bmv2 proto is used to initially parse the JSON format, which is then
+// processed by our tool and transformed into the IR.
+// The IR does not cover all the features of P4 programs.
+// Specifically, we intentionally ignore parsers, de-parsers, stacks and unions,
+// checksum calculations, and extern definitions.
+// Additionally, some types of expressions and statements are unsupported, these
+// are described in the relevant sections of this file.
+// See the bmv2 JSON format reference for more information:
+// https://github.com/p4lang/behavioral-model/blob/master/docs/JSON_format.md
+
+syntax = "proto3";
+
+package p4_symbolic.ir;
+
+import "p4_pdpi/ir.proto";
+import "p4_symbolic/bmv2/bmv2.proto";
+
+// Overall structure of a p4 program.
+message P4Program {
+  // Header definitions, keyed by the header type name (e.g. standard_metadata).
+  map<string, HeaderType> headers = 1;
+  // Action definitions and implementations, keyed by action name.
+  map<string, Action> actions = 2;
+  // Table definitions and implementations, keyed by table name.
+  map<string, Table> tables = 3;
+  // The name of the initial table.
+  string initial_table = 6;
+}
+
+// A header definition. Includes the programmer assigned name,
+// the header id, and the fields in the heade.
+message HeaderType {
+  string name = 1;
+  int32 id = 2;
+  // Maps a field name to its definition.
+  map<string, HeaderField> fields = 3;
+}
+
+// A P4 header type definition.
+// Includes its name, bitwidth (type), and sign flag, and the
+// enclosing header type name that defines this field.
+message HeaderField {
+  string name = 1;
+  int32 bitwidth = 2;
+  bool signed = 3;
+  string header_type = 4;
+}
+
+// Overall action structure, combining definition with implementation.
+message Action {
+  // This contains the name, id, and parameters, among other things.
+  // Essentially, this is the action signature.
+  pdpi.ir.IrActionDefinition action_definition = 1;
+  // Action implementation, essentially its body.
+  ActionImplementation action_implementation = 2;
+}
+
+// Action implementation details.
+message ActionImplementation {
+  // List of statements in action body in sequence.
+  repeated Statement action_body = 1;
+}
+
+// Overall table structure, combining definition with implementation.
+message Table {
```

Comment:
I have not noticed this earlier, but bmv2 actually converts all kinds of things into tables, not just P4 tables. So there won't be a table definition from P4Info for some things. That seems like a problem, not sure what the solution is.